cmd/noms/diff/diff.go

```diff
@@ -5,121 +5,136 @@
-
-const (
-	ADD = "+   "
-	DEL = "-   "
 )
 
 type (
 	diffFunc     func(changeChan chan<- types.ValueChanged, stopChan <-chan struct{})
-	lineFunc     func(w io.Writer, op prefixOp, key, val types.Value) error
 	pathPartFunc func(v types.Value) types.PathPart
 	valueFunc    func(k types.Value) types.Value
 )
 
-type diffWriter struct {
-	w         io.Writer
-	leftRight bool
+// Difference represents a "diff" between two Noms graphs.
+type Difference struct {
+	// Path to the Value that has changed
+	Path types.Path
+	// ChangeType indicates the type of diff: modified, added, deleted
+	ChangeType types.DiffChangeType
+	// OldValue is Value before the change, can be nil if Value was added
+	OldValue types.Value
+	// NewValue is Value after the change, can be nil if Value was removed
+	NewValue types.Value
 }
 
-func (w diffWriter) Write(p []byte) (n int, err error) {
-	return w.w.Write(p)
-}
+type (
+	// Differences is a slice of Difference
+	Differences []Difference
 
-func shouldDescend(v1, v2 types.Value) bool {
-	kind := v1.Type().Kind()
-	return !types.IsPrimitiveKind(kind) && kind == v2.Type().Kind() && kind != types.RefKind
-}
+	// DiffReceiveChan is used by caller to process diffs
+	DiffReceiveChan <-chan Difference
 
-// Diff writes the diff from |v1| to |v2| to |w|.
-// If |leftRight| is true then the left-right diff is used for ordered sequences - see Diff vs DiffLeftRight in Set and Map.
-func Diff(w io.Writer, v1, v2 types.Value, leftRight bool) error {
-	if v1.Equals(v2) {
-		return nil
-	}
+	// StopSendChan is used by the caller to signal that Diff() should stop processing
+	StopSendChan chan<- struct{}
 
-	dw := diffWriter{w, leftRight}
+	stopReceiveChan <-chan struct{}
+	diffSendChan    chan<- Difference
+)
 
-	if !shouldDescend(v1, v2) {
-		line(dw, DEL, nil, v1)
-		return line(dw, ADD, nil, v2)
-	}
+// Diff traverses two graphs simultaneously looking for differences. It returns
+// two channels: a DiffReceiveChan that the caller can use to iterate over the
+// diffs in the graph and a StopSendChanel that a caller can use to signal the
+// Diff function to stop processing.
+// Diff returns the Differences in depth-first first order. A 'diff' is defined
+// as one of the following conditions:
+// * a Value is Added or Removed from a node in the graph
+// * the type of a Value has changed in the graph
+// * a primitive (i.e. Bool, Number, String, Ref or Blob) Value has changed
+//
+// A Difference is not returned when a non-primitive value has been modified. For
+// example, a struct field has been changed from one Value of type Employee to
+// another. Those modifications are accounted for by the Differences described
+// above at a lower point in the graph.
+//
+// If |leftRight| is true then the left-right diff is used for ordered sequences
+// - see Diff vs DiffLeftRight in Set and Map.
+//
+func Diff(v1, v2 types.Value, leftRight bool) (DiffReceiveChan, StopSendChan) {
+	diffChan := make(chan Difference, 256)
```

Comment:
And passing channels into this function rather than creating them lets the caller decide the buffer size. 256 is probably unnecessarily large.